### ISSUE TITLE
Add except: ["after-closing-brace"] to block-closing-brace-empty-line-before

### ISSUE THREAD

**Issue by thomasjbradley** [stylelint/stylelint#2090]
> Describe the issue. Is it a bug or a feature request (new rule, new option, etc.)?

Possibly a bug, but most likely a feature request.

I'd like to be able to force an empty line at the end of a media query—we can do it at the top of a media query, but I can't seem to find how to do it at the bottom.

```css
@media print {

  a {
    color: red;
  }
            /* <----- Force this empty line here */
}
```

> Which rule, if any, is this issue related to?

I see the rule: `rule-nested-empty-line-before` which will force a new line at the top of the media query.

And I see the rule: `block-closing-brace-empty-line-before` which will force a new line at the end of the block, but that would also force a new line after `color: red`.

> What CSS is needed to reproduce this issue?

*This CSS should pass:*

```css
@media print {

  a {
    color: red;
  }

}
```

*This CSS should fail:*

```css
@media print {

  a {
    color: red;
  }
}
```

> Which version of stylelint are you using?

Currently using `7.5.0`.

> How are you running stylelint: CLI, PostCSS plugin, Node API?

Node API.

> Does your issue relate to non-standard syntax (e.g. SCSS, nesting, etc.)?

Nope.

> What did you expect to happen?

I'd like to get a warning message if there isn’t an empty line at the end of a media query.

> What actually happened (e.g. what warnings or errors you are getting)?

Both empty lines and non empty lines pass Stylelint tests.

I feel like an older version used to enforce this, but looking through past commits of my Stylelint config I don't see anything that jumps out at me as being “the rule” that enforced that requirement.

**Comment by ntwb** [comment]
Thanks for using our issue template and the detailed description @thomasjbradley 👍 

p.s. Originally via [this](https://stackoverflow.com/questions/40593334/stylelint-force-empty-line-at-bottom-of-nested-block) Stack Overflow issue.

**Comment by jeddy3** [comment]
I think we can achieve this with the addition of an `exceptAtRules: []` option to `block-closing-brace-empty-line-before`.

There is a precedence for `*-empty-line-before` rules accepting _user defined_ options in [at-rule-empty-line-before](https://github.com/stylelint/stylelint/tree/master/src/rules/at-rule-empty-line-before#ignoreatrules-array-of-at-rules). And one for `block-closing-brace-*` rules in [block-closing-brace-newline-after](https://github.com/stylelint/stylelint/tree/master/src/rules/block-closing-brace-newline-after#ignoreatrules-regex-non-regex).

This would be _the first instance_ of a _user defined_ `except*: []` option though, but I can't foresee any issues with such an option type.

So, the following config:

```json
"block-closing-brace-empty-line-before": ["never", { "exceptAtRules": ["media"] }]
```

Would enforce this:

```css
@media print {

  a {
    color: red;
  }

}
```

Having an _user defined_ option, rather than just the _keyword_ one of `except: ["media"]`, provides flexibility for applying a similar code style to other at-rules e.g. `@supports`:

```json
"block-closing-brace-empty-line-before": ["never", { "exceptAtRules": ["media", "supports"] }]
```

```css
@media print {

  a {
    color: red;
  }

}

@supports (display:flex) {

  a {
    display: flex;
  }

}
```

Unless there are objections, I'll label this up as "help wanted" and point @thomasjbradley in the direction of how to contribute this option.

**Comment by davidtheclark** [comment]
@jeddy3 What if someone wants to use the exception for *all* at-rules?

**Comment by jeddy3** [comment]
> What if someone wants to use the exception for all at-rules?

Oops, I hadn't thought about that. 

@thomasjbradley Is your code style to have an empty line before the closing brace of _just_ `@media` or [all at-rules](https://developer.mozilla.org/en/docs/Web/CSS/At-rule) e.g. `@support`, `@page`, `@font-face`, or do you have one code style for at-rules that nest rule-sets (e.g. `@media`, `@keyframes` etc), but not for those that only contain declarations e.g. `@font-face`?

That last one makes me think that having the flexibility to specify exactly which at-rules are exceptions is desirable, but I'm 100% sure now.

**Comment by thomasjbradley** [comment]
For me it would only be nested rulesets. So `@supports`, `@keyframes`, `@media` — but not `@font-face`.

**Comment by jeddy3** [comment]
@davidtheclark So it sounds like could have an `except: ["nesting-at-rules"]` keyword option, but I'm not in favour of that because it doesn't account for custom at-rules e.g. 

```css
@my-custom-nesting-at-rule {

  a {}

}
```

The `exceptAtRules: []` option does require more configuration, but it feels necessary for this flexibility.

**Comment by davidtheclark** [comment]
> it doesn't account for custom at-rules

Why not? If we're just checking that the at-rule contains rules, that would apply just as much to custom at-rules as standard ones, right?

**Comment by jeddy3** [comment]
I realised the same thing on the train. I originally thought we'd have to keep a list of standard nesting at-rules, but now, like you, I see that we can just detect if the at-rule contains rules.

`except: ["nesting-at-rules"]` works for me.

**Comment by jeddy3** [comment]
I'll label this issue up accordingly.

@thomasjbradley Would you [like to contribute](https://github.com/stylelint/stylelint/blob/master/docs/developer-guide/rules.md#adding-options-to-existing-rules) this option?  [`rule-non-nested-empty-line-before`](https://github.com/stylelint/stylelint/blob/master/src/rules/rule-non-nested-empty-line-before/index.js#L63-L77) has some examples of how such an `except` option might work. In this instance, you'll want to check if the `atRule` node contains any `atRule` or `rule` children and then invert the primary option if it does.

**Comment by jeddy3** [comment]
> For me it would only be nested rulesets

In line with https://github.com/stylelint/stylelint/issues/1663, I'm pretty sure this option should be `"after-closing-brace"`.

**Comment by jonathantneal** [comment]
I would like to define an exception for `block-closing-brace-newline-after`. Is this possible now, or am I just seconding this request? My use-case is:

```json
{
  "block-closing-brace-newline-after": ["always", { "exceptAtRules": ["else"] }]
}
```

```css
@media (width < 500) {
	display: none;
} @else {
	display: block;
}
```

Citations:
https://tabatkins.github.io/specs/css-when-else/#else-rule
https://github.com/jonathantneal/postcss-at-else

**Comment by jeddy3** [comment]
@jonathantneal It is possible now using a couple of rules from [stylelint-scss](https://github.com/kristerkari/stylelint-scss) (a plugin pack for SCSS-_like_ constructs). Those guys [wrote up an example](https://github.com/kristerkari/stylelint-scss/blob/master/docs/examples/if-else.md) on how to do it.

**Comment by ArmorDarks** [comment]
Maybe we need even more generic approach, which could be used almost with any Stylelint rule?

For instance, we can determinate interface, which will allow to target within ignore _any_ recognized by Stylelint CSS entity and whenever needed, to specify which exactly instance of entity should it be?

For instance, it could be something like

```yaml
ignore:
- property
- at-rule:
  - 'media'
- ruleset
- whatever_else_Stylelint_recognizes
```

**Comment by ArmorDarks** [comment]
It is kinda strange when here https://github.com/stylelint/stylelint/issues/2090#issuecomment-262533548 and here https://github.com/stylelint/stylelint/issues/2090#issuecomment-316424488 we were talking about some nicer, more generic and broad interface, and than issue suddenly closed by slightly PR, which solves some problems, but obviously doesn't work as a solution for author's original issue with enforcing empty line _only_ in at rules, while the PR allows doing so only globally
